### ISSUE TITLE
The results are different every time. A memory leak?

### ISSUE THREAD

**Issue by mwwzbinf** [SurfaceMan/surface_match#11]
![image](https://github.com/SurfaceMan/surface_match/assets/133193722/17969ae1-0ab0-41e7-9b73-092e93a6016b)
![image](https://github.com/SurfaceMan/surface_match/assets/133193722/ef9d4ead-8ac0-413b-b0e4-8e02ced1996d)
![image](https://github.com/SurfaceMan/surface_match/assets/133193722/d2ed45e5-2b93-4011-a0f2-333eecdf1048)

I successfully compiled the program, but every time the result is different, or no result, or wrong result, or correct result, and accompanied by an error pop-up window, I do not know why? Is it a problem with openmp? I used my own model and scene, and the model is parasaurolophus_6700.ply in opencv, Can you help me?

**Comment by SurfaceMan** [comment]
1.1 I have no ideas about the error before. If you have any progress, please let me known.
1.2 Is OpenMP problem? You can try comment OpenMP pragma with //.
1.3 I gauss may std::vector(score & pose) construct(ppf.dll) and destroy(test_ppf.exe) at different place cause this error.
2.1 About result, sometimes got a strange pose before icp, there may need sometime to analyze it.

**Comment by mwwzbinf** [comment]
The memory leak is indeed due to the third point you mentioned, I am understanding your code, is there a blog that can understand it better?

**Comment by SurfaceMan** [comment]
> The memory leak is indeed due to the third point you mentioned, I am understanding your code, is there a blog that can understand it better?

try reference this [issue](https://github.com/SurfaceMan/surface_match/issues/6)

**Comment by mwwzbinf** [comment]
sampleMesh has a significant effect on the results, Perhaps a fixed sampling method would be better

**Comment by SurfaceMan** [comment]
> sampleMesh has a significant effect on the results, Perhaps a fixed sampling method would be better

You mean sample process lead to no result sometime?

**Comment by mwwzbinf** [comment]
yes, I use parasaurolophus_6700.ply and rs1_normals.ply, you can try
![image](https://github.com/SurfaceMan/surface_match/assets/133193722/879d7bc2-6676-4321-b65c-21e0e63a361c)

**Comment by SurfaceMan** [comment]
I think the main reason is a bug of cluster method which cause this error. the cluster2 method only keep max votes result which maybe not the true result.

**Comment by mwwzbinf** [comment]
I have some questions
1.Why not use the green part instead of the red part? What is the purpose of the calculation in red part?
![image](https://github.com/SurfaceMan/surface_match/assets/133193722/725886a4-0f80-4ed2-b879-88a88ea555b5)
2.Why add a pi and subtract a pi? I think we can just delete them.
![image](https://github.com/SurfaceMan/surface_match/assets/133193722/6f8d88d6-c5a0-43e5-8579-ca1fd9601699)
![image](https://github.com/SurfaceMan/surface_match/assets/133193722/72a419c3-2cae-409d-bcac-a02941b836c9)

**Comment by SurfaceMan** [comment]
1.1 I also not understand yet. The code based opencv surface_match.
1.2  I tested replace with "float alpha = atan2 (mpt (2), mpt (1));", it will lost many result.
2. alpha angle range is(-pi, pi), the difference of model angle and scene angle is (-2pi, 2pi), we should normalize it to (0, 2pi) then map it to angle index (0, maxAngleIndex)

**Comment by mwwzbinf** [comment]
one question, the code in samplePointCloud, due to the use of openmp, two points that are very close together (less than the radius) are preserved, Whether there is such a chance?
![image](https://github.com/SurfaceMan/surface_match/assets/133193722/d79464d5-2945-47eb-9ac2-1f4c5377fda3)

**Comment by SurfaceMan** [comment]
Yes, the multithread lead to random sample result. 
Because of the big difference between thread count and point count, it looks link uniform.

**Comment by mwwzbinf** [comment]
Thank you very much. One more question. Is there any article you can refer to about ICP?

**Comment by SurfaceMan** [comment]
> Thank you very much. One more question. Is there any article you can refer to about ICP?

please see first of [TODO](https://github.com/SurfaceMan/surface_match/blob/master/TODO.md) list

**Comment by XiaXingLuo** [comment]
> I think the main reason is a bug of cluster method which cause this error. the cluster2 method only keep max votes result which maybe not the true result.

How to deal with the problems about not converage?
```
after cluster has items: 2
sparsePoseRefinement not converge 2
sparsePoseRefinement not converge 2
```
